### PR TITLE
KEYCLOAK-13916 Refactor SAML metadata generation to use the SAMLMetadataWriter class

### PR DESCRIPTION
**Pull request by lscorcia**
Changes the current IdP/SP metadata generation logic, based on string manipulations, to use the more robust SAMLMetadataWriter class. The generated XML is functionally the same as before. 

Small notes:
* I had to add a couple of constants to the JBossSAMLURIConstants class. Are those really useful? Does Keycloak client really support SAML1/SECEXT protocols?
* I had to fix some minor errors in the SAMLMetadataWriter class, i.e.:
  a. Attribute protocolSupportEnumeration in SPSSODescriptor: was rendering only the first protocol of the supplied list
  b. SPSSODescriptor did not output KeyDescriptors
  c. IDPSSODescriptor-generated files were not compliant with the SAML Metadata XSD (the NameIDFormat tag must be placed after the SingleLogout elements)
* I had to change the value of the constant FALLBACK_ERROR_URL_STRING to make it a valid URI (replaced spaces with underscores)
* The previous generated XMLs were indented - after this PR they no longer are. I haven't been able to find a way to set the XMLStreamWriter class to pretty print. Suggestions welcome!
* The tests have been updated and are all passing (after pulling in https://github.com/keycloak/keycloak/pull/6993).

### CODE REVIEW

**Comment by lscorcia** [comment]
I have no idea why the automatic tests failed, the tests complete successfully if I run `mvn -Dtest=IdentityProviderTest -DfailIfNoTests=false package`, even after merging latest master into my branch. Am I running the tests incorrectly?

**Comment by lscorcia** [comment]
I fixed the failing tests with a simple check, but the right fix would probably to correctly import the `NameIDFormat` tag in the `SamlIdentityProviderFactory.parseConfig` class. Anyway I felt like that would deserve a separate pull request. Please, let me know if it's good enough to go in!

**Comment by lscorcia** [comment]
Rebased to current and squashed commits.

File: services/src/main/java/org/keycloak/protocol/saml/IDPMetadataDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.saml;
+
+import org.keycloak.dom.saml.v2.metadata.*;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/IDPMetadataDescriptor.java:20]: Please avoid star imports

File: services/src/main/java/org/keycloak/protocol/saml/installation/SamlSPDescriptorClientInstallation.java [hunk 2]
```diff
@@ -28,45 +30,55 @@
...
+                binding = JBossSAMLURIConstants.SAML_HTTP_POST_BINDING.getUri();
+            } else { //redirect binding
+                assertionUrl = client.getAttribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_REDIRECT_ATTRIBUTE);
+                logoutUrl = client.getAttribute(SamlProtocol.SAML_SINGLE_LOGOUT_SERVICE_URL_REDIRECT_ATTRIBUTE);
+                binding = JBossSAMLURIConstants.SAML_HTTP_REDIRECT_BINDING.getUri();
+            }
+            if (assertionUrl == null || assertionUrl.trim().isEmpty()) assertionUrl = client.getManagementUrl();
+            if (assertionUrl == null || assertionUrl.trim().isEmpty()) assertionUrl = FALLBACK_ERROR_URL_STRING;
+            if (logoutUrl == null || assertionUrl.trim().isEmpty()) logoutUrl = client.getManagementUrl();
+            if (logoutUrl == null || assertionUrl.trim().isEmpty()) logoutUrl = FALLBACK_ERROR_URL_STRING;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/installation/SamlSPDescriptorClientInstallation.java:70]:
```suggestion
            if (logoutUrl == null || logoutUrl.trim().isEmpty()) logoutUrl = client.getManagementUrl();
            if (logoutUrl == null || logoutUrl.trim().isEmpty()) logoutUrl = FALLBACK_ERROR_URL_STRING;
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you very much for your contribution! Looking at the code, the changes look good to me, nice work!

> I had to add a couple of constants to the JBossSAMLURIConstants class. Are those really useful? Does Keycloak client really support SAML1/SECEXT protocols?

You are correct, neither of them is needed, please remove them.

> The previous generated XMLs were indented - after this PR they no longer are. I haven't been able to find a way to set the XMLStreamWriter class to pretty print. Suggestions welcome!

No worries, indentation is not a concern at all.

Apart from this, only a few nits and the PR is ready to go.

File: services/src/main/java/org/keycloak/protocol/saml/IDPMetadataDescriptor.java (hunk 1 above)
Comment (lscorcia) [reply services/src/main/java/org/keycloak/protocol/saml/IDPMetadataDescriptor.java]: Thanks, done!

File: services/src/main/java/org/keycloak/protocol/saml/installation/SamlSPDescriptorClientInstallation.java (hunk 2 above)
Comment (lscorcia) [reply services/src/main/java/org/keycloak/protocol/saml/installation/SamlSPDescriptorClientInstallation.java]: Nice catch!

**Comment by lscorcia** [comment]
Applied code review suggestions, squashed commits and rebased to current.

**Review by hmlnarik — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**